### PR TITLE
bpo-46433: _PyType_GetModuleByDef: handle static types in MRO

### PR DESCRIPTION
**Pull request by encukou**
https://bugs.python.org/issue46433

Automerge-Triggered-By: GH:encukou

### CODE REVIEW

File: Modules/_testmultiphase.c [hunk 1]
```diff
@@ -153,6 +155,24 @@ _testmultiphase_StateAccessType_get_defining_module_impl(StateAccessTypeObject *
...
+Test that result of _PyType_GetModuleByDef with a bad def is NULL.
+[clinic start generated code]*/
+
+static PyObject *
+_testmultiphase_StateAccessType_getmodulebydef_bad_def_impl(StateAccessTypeObject *self,
+                                                            PyTypeObject *cls)
+/*[clinic end generated code: output=64509074dfcdbd31 input=edaff09aa4788204]*/
+{
+    _PyType_GetModuleByDef(Py_TYPE(self), &def_nonmodule);  // should raise
+    assert(PyErr_Occurred);
```
Comment (kumaraditya303) [inline Modules/_testmultiphase.c]:
```suggestion
    assert(PyErr_Occurred());
```

**Review by kumaraditya303 — CHANGES_REQUESTED** [review]

File: Modules/_testmultiphase.c (hunk 1 above)
Comment (encukou) [reply Modules/_testmultiphase.c]: Thanks for the catch!

**Comment by encukou** [comment]
Ah, turns out I won't be able to optimize this further. `_PyType_GetModuleByDef` is currently called on static types, so the flag needs to be checked.
Removing the calls to `_PyType_GetModuleByDef` on static types isn't feasible IMO: it's too easy to miss the requirement in the future.

@vstinner, since you did the optimization, could you look at this code? Did I miss something performance-related?

File: Objects/typeobject.c [hunk 2]
```diff
@@ -3747,11 +3747,11 @@ _PyType_GetModuleByDef(PyTypeObject *type, struct PyModuleDef *def)
...
     for (Py_ssize_t i = 0; i < n; i++) {
         PyObject *super = PyTuple_GET_ITEM(mro, i);
-        // _PyType_GetModuleByDef() must only be called on a heap type created
-        // by PyType_FromModuleAndSpec() or on its subclasses.
-        // type_ready_mro() ensures that a static type cannot inherit from a
-        // heap type.
-        assert(_PyType_HasFeature(type, Py_TPFLAGS_HEAPTYPE));
+        if(!_PyType_HasFeature((PyTypeObject *)super, Py_TPFLAGS_HEAPTYPE)) {
+            // Static types in the MRO need to be skipped
+            i++;
```
Comment (vstinner) [inline Objects/typeobject.c]: `i++` looks wrong: continue does run the for loop `i++` part, no?

Comment (encukou) [reply Objects/typeobject.c]: You're right. I switched between `for` and `while` too much and missed this. Thank you!

File: Lib/test/test_capi.py [hunk 3]
```diff
@@ -1068,6 +1068,18 @@ def test_state_access(self):
                 with self.assertRaises(TypeError):
                     increment_count(1, 2, 3)
 
+    def test_get_module_bad_def(self):
```
Comment (vstinner) [inline Lib/test/test_capi.py:1071]: You may mention "bpo-46433" in a comment, here and in test_get_module_static_in_mro().

**Review by vstinner — APPROVED** [review]
LGTM.

**Comment by encukou** [comment]
Thanks for the review!

**Pull request merged**